### PR TITLE
Issue #10932: remove MainPowerTest

### PR DESCRIPTION
**Pull request by pbludov**
Part of #10932 

It is no more possible to mock `System.exit` nor replace `ordinal` field of an `enum`.
For the fields `System.err` and `System.out` there is a workaround. We can replace them with a custom stream.

Limitation:
```
✔ ~/java/github/romani/checkstyle [pbludov/issue-10932-main L|✚ 1] 
$ git diff
diff --git a/src/main/java/com/puppycrawl/tools/checkstyle/Main.java b/src/main/java/com/puppycrawl/tools/checkstyle/Main.java
index 8d5a261..ae0724d 100644
--- a/src/main/java/com/puppycrawl/tools/checkstyle/Main.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/Main.java
@@ -149,7 +149,9 @@ public final class Main {
                 System.err.println(errorCounterViolation.getViolation());
             }
         }
-        ExitHelper.exit(exitStatus);
+        if (exitStatus != 0) {
+            ExitHelper.exit(exitStatus);
+        }
     }
 
     /**
@@ -878,9 +880,7 @@ public final class Main {
          * @noinspection CallToSystemExit
          */
         private static void exit(int code) {
-            if (code != 0) {
-                System.exit(code);
-            }
+            System.exit(code);
         }
 
     }
```

image
So new class is created, to hide this uncoverage.

### CODE REVIEW

File: pom.xml
Comment (rnveach) [inline pom.xml:927]: @romani ping
So the only way is to drop line coverage for the class?

Comment (pbludov) [reply pom.xml:927]: @rnveach some workarounds are possible. We can keep the coverage at 100% for the `Main` class.
By moving the call to `System.exit` to some other class (this class will be not 100% covered).
Or we can mark this code with special annotation to turn JaCoCo coverage off for one method.
Both ways are tricky, it is better to count the coverage as is.

Comment (nmancus1) [reply pom.xml:927]: > Or we can mark this code with special annotation to turn JaCoCo coverage off for one method.

The only way I see to do this is using `@generated` annotation, this is not ideal.

> By moving the call to System.exit to some other class (this class will be not 100% covered).

This seems like a hack to me.

I think it is better that we stay with coverage specification in `pom.xml`, it is the probably the best option out of the three given.

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java
Comment (nmancus1) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java:408]: What happens if `try` fails?

File: src/test/java/com/puppycrawl/tools/checkstyle/gui/MainFrameModelTest.java
Comment (nmancus1) [inline src/test/java/com/puppycrawl/tools/checkstyle/gui/MainFrameModelTest.java:191]: I feel that each assertion should have it's own test method that describes our assertion. The method name `testShouldAcceptFile` doesn't accurately describe each assertion here.

**Review by nmancus1 — CHANGES_REQUESTED** [review]
Thoughts:

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java
Comment (pbludov) [reply src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java:408]: try-with-resource is used here  to avoid "Closable resource not closed" violation.

File: src/test/java/com/puppycrawl/tools/checkstyle/gui/MainFrameModelTest.java
Comment (pbludov) [reply src/test/java/com/puppycrawl/tools/checkstyle/gui/MainFrameModelTest.java]: Done. The whole test class should be rewritten, but this is another story: #9142

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java
Comment (strkkk) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java:421]: minor: it seems weird that identation of last 2 assertions are different from first one

Comment (pbludov) [reply src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: done

File: pom.xml
Comment (nmancus1) [reply pom.xml:927]: On second thought, could we cover `System.exit()` call by execution of `Main#main` with  `TestUtil#getResultWithLimitedResources` ?

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java
Comment (nmancus1) [reply src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java:408]: Resolved.

File: src/test/java/com/puppycrawl/tools/checkstyle/gui/MainFrameModelTest.java
Comment (nmancus1) [reply src/test/java/com/puppycrawl/tools/checkstyle/gui/MainFrameModelTest.java]: Resolved.

**Review by nmancus1 — APPROVED** [review]

**Review by strkkk — APPROVED** [review]

**Comment by rnveach** [comment]
@romani https://github.com/checkstyle/checkstyle/pull/10951#discussion_r748880757

File: pom.xml
Comment (pbludov) [reply pom.xml:927]: > On second thought, could we cover `System.exit()` call by execution of `Main#main` with `TestUtil#getResultWithLimitedResources` ?

`System.exit()` just forces VM to quit. No coverage probes will be touched.
This can be hacked with a custom `SecurityManager`, but in this case this method will throw an exception.
The problem is that
```java
System.exit(1); // Coverage collected for this  line
// Coverage NOT COLLECTED for this  line 
} // Coverage collected for this  line, the program ends here
```

The workaround we using for JDK < 16 is about to simply mock `System.exit`. But now this is not possible.

**Comment by romani** [comment]
@pbludov , please answer at https://github.com/checkstyle/checkstyle/issues/10161

**Comment by pbludov** [comment]
> @pbludov , please answer at #10161

done

File: pom.xml
Comment (pbludov) [reply pom.xml]: Finally solved with a helper inner class. Now JaCoCo coverage for the `Main` class is 100%

File: config/pmd.xml
Comment (pbludov) [inline config/pmd.xml:207]: The rule was renamed in PMD 4.1

**Comment by pbludov** [comment]
GitHub, rebase

File: src/main/java/com/puppycrawl/tools/checkstyle/Main.java
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/Main.java:882]: Can we keep IF condition out of Helper class ?

As we keep this class only to make Jacoo happy 
So better to keep only problematic  statement.

**Review by romani — CHANGES_REQUESTED** [review]
Item:

File: src/main/java/com/puppycrawl/tools/checkstyle/Main.java
Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/Main.java:882]: No. In this case the coverage of the `Main` class will be 99% because the `ExitHelper.exit()` will never return:

```java
 if (statusCode != 0) {
  ExiitHelper.exit(statusCode);
  // This line will be marked as uncovered
}
```

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/Main.java:882]: you mean coverage will drop if we do:
```
        private static void exit(int code) {
                System.exit(code);
        } 
```

if it drops it means problem is in IF condition coverage.

what do you mean by `will never return` ? it will never finish to execute ? or it will not actually exit from java process ?

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/Main.java:882]:
```
✔ ~/java/github/romani/checkstyle [pbludov/issue-10932-main L|✚ 1] 
05:55 $ git diff
diff --git a/src/main/java/com/puppycrawl/tools/checkstyle/Main.java b/src/main/java/com/puppycrawl/tools/checkstyle/Main.java
index 8d5a261..ae0724d 100644
--- a/src/main/java/com/puppycrawl/tools/checkstyle/Main.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/Main.java
@@ -149,7 +149,9 @@ public final class Main {
                 System.err.println(errorCounterViolation.getViolation());
             }
         }
-        ExitHelper.exit(exitStatus);
+        if (exitStatus != 0) {
+            ExitHelper.exit(exitStatus);
+        }
     }
 
     /**
@@ -878,9 +880,7 @@ public final class Main {
          * @noinspection CallToSystemExit
          */
         private static void exit(int code) {
-            if (code != 0) {
-                System.exit(code);
-            }
+            System.exit(code);
         }
 
     }
```

image

**Review by romani — APPROVED** [review]
ok to merge.

I changed comment above ExitHelper

**Comment by romani** [comment]
@pbludov , better to wait for review from @rnveach as he has comments on this PR

**Comment by pbludov** [comment]
> @pbludov , better to wait for review from @rnveach as he has comments on this PR

Sorry. @rnveach  please take a look. I will create a new PR if you have any notes.

**Comment by rnveach** [comment]
My main concern was that agreement was made in dropping Jacoco coverage which was added to the pom. https://github.com/checkstyle/checkstyle/pull/10951/files#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8

I was curious at https://github.com/checkstyle/checkstyle/pull/10951#pullrequestreview-817134894 why we didn't catch this in earlier PMD upgrades or recent versions. Are we sure the violation is still valid or is just both names are valid in PMD?

**Comment by pbludov** [comment]
> I was curious at [#10951 (review)](https://github.com/checkstyle/checkstyle/pull/10951#pullrequestreview-817134894) why we didn't catch this in earlier PMD upgrades or recent versions. Are we sure the violation is still valid or is just both names are valid in PMD?

Both are valid. `DoNotCallSystemExit` is just an alias to `DoNotTerminateVM`:
https://github.com/pmd/pmd/blob/master/pmd-java/src/main/resources/category/java/errorprone.xml#L1351

Note that this rule is ignored for `static void main` method:
https://github.com/pmd/pmd/blob/master/pmd-java/src/main/resources/category/java/errorprone.xml#L1435

After moving this call to a proxy class, this violation began to be reported.

**Pull request merged**
